### ISSUE TITLE
ppc64/ppc64le builds fail with "Error: number '1e-300' is not representable"

### ISSUE THREAD

**Issue by kalev** [ldc-developers/ldc#2046]
1.2.0 beta1 fails to build on Fedora ppc64 and ppc64le builders with:

```
[ 13%] Generating src/core/stdc/fenv.o
cd /builddir/build/BUILD/ldc-1.2.0-beta1-src && /builddir/build/BUILD/ldc-1.2.0-beta1-src/build/bin/ldc2 --output-o -c -I/builddir/build/BUILD/ldc-1.2.0-beta1-src/runtime/druntime/src -I/builddir/build/BUILD/ldc-1.2.0-beta1-src/runtime/druntime/src/gc /builddir/build/BUILD/ldc-1.2.0-beta1-src/runtime/druntime/src/core/stdc/fenv.d -of/builddir/build/BUILD/ldc-1.2.0-beta1-src/build/runtime/src/core/stdc/fenv.o -w -relocation-model=pic -O3 -release -d-version=Shared
/builddir/build/BUILD/ldc-1.2.0-beta1-src/runtime/druntime/src/core/stdc/fenv.d(399): Error: number '1e-300' is not representable
/builddir/build/BUILD/ldc-1.2.0-beta1-src/runtime/druntime/src/core/stdc/fenv.d(400): Error: number '1e-300' is not representable
make[2]: *** [runtime/CMakeFiles/druntime-ldc.dir/build.make:325: runtime/src/core/stdc/fenv.o] Error 1
```

Full build logs available at https://kojipkgs.fedoraproject.org//work/tasks/922/18460922/build.log (short lived link)

**Comment by kinke** [comment]
This originates from [here](https://github.com/ldc-developers/ldc/blob/5c5831a32f56dbc2e3aba142d7b3d77bebf587d1/ddmd/lexer.d#L2089-L2135) and should be the case if the C runtime's `strtold("1e300", null)` and/or `strtod("1e300", null)` leave behind an `errno == ERANGE` (setting `*isOutOfRange` [here](https://github.com/ldc-developers/ldc/blob/master/ddmd/root/ctfloat.d#L121-L138) or returning true [here](https://github.com/ldc-developers/ldc/blob/master/ddmd/root/port.d#L91-L112)).
Both options seem odd as decimal exponents up to 307 shouldn't be a problem for double precision. Maybe the `strtold()` signature is wrong in assuming D `real` (double precision for LDC on PPC) is C `long double` (that strange doubledouble?).

**Comment by kinke** [comment]
Btw, this over-/underflow check wasn't in 2.071 if I recall correctly, so this would have gone unnoticed during compilation with earlier LDC versions, but a bunch of unittests would have failed if the host compiler's C runtime really fails to parse subnormal float/double literals. [`real` literals are an exception and aren't checked, otherwise the compiler's lexer wouldn't be able to handle literals in explicit `version(...)` or `static if` blocks.]

**Comment by kalev** [comment]
Right, we haven't been running tests during the builds so far. I guess we should?

**Comment by kinke** [comment]
The results on ppc64 should be rather bad due to incomplete big-endian support in Phobos, see [here](https://github.com/ldc-developers/ldc/issues/1909#issuecomment-264558022). AArch64 is seriously incomplete (varargs, quad-precision math, ...). As long as the tests are allowed to fail, it would be very interesting to have test results on those platforms. ppc64le might even make it through ;), ARM should pass afaik.

**Comment by kinke** [comment]
Oh and @kalev: Any chance you can test something like this?

```C++
#include <cerrno>
#include <cstdio>
#include <cstdlib>

void main()
{
    errno = 0; auto r = strtold("1e-300", nullptr); printf("%La, %d\n", r, errno == ERANGE ? 1 : 0);
    errno = 0; auto d = strtod("1e-300", nullptr); printf("%a, %d\n", d, errno == ERANGE ? 1 : 0);
}
```

**Comment by kalev** [comment]
Sure, this is what I get on a ppc64le machine:
```
0x1.56e1fc2f8f358d94db78p-997, 1
0x1.56e1fc2f8f359p-997, 0
```

**Comment by kinke** [comment]
Thanks, very helpful. I'd have bet that strtod() fails, but it's `strtold()` ending up with ERANGE (but the parsed value is fine). According to the [Linux man page](https://linux.die.net/man/3/strtold), that's invalid behavior:

> If the correct value would cause overflow, plus or minus HUGE_VAL (HUGE_VALF, HUGE_VALL) is returned (according to the sign of the value), and ERANGE is stored in errno. If the correct value would cause underflow, zero is returned and ERANGE is stored in errno.

In #1317, I used LLVM to parse floating-point literals. If I recall correctly, the problem with that was that it dies with an assertion if the string is invalid, so the string would need to be checked beforehand...

**Comment by kinke** [comment]
Should be fixed by #2054.

**Comment by kalev** [comment]
Thanks @kinke! I'll give it a try when the next beta tarball is out.

**Comment by kalev** [comment]
Confirmed as fixed in 1.2.0 beta2. Thanks!

**Comment by kinke** [comment]
Great, thanks for reporting & confirming!
